### PR TITLE
Arnej/redo nn distance for new cell types 2 

### PR DESCRIPTION
**Pull request by arnej27959**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@havardpe please review
@geirst please also take a look

### CODE REVIEW

File: searchlib/src/vespa/searchlib/tensor/distance_function.h [hunk 1]
```diff
@@ -15,10 +16,20 @@ namespace search::tensor {
  * The actual implementation must know which type the vectors are.
  */
 class DistanceFunction {
+protected:
+    vespalib::eval::CellType expect_cell_type;
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/tensor/distance_function.h:20]: Consider using '_' prefix for member variable.

File: searchlib/src/tests/tensor/hnsw_index/hnsw_index_test.cpp [hunk 2]
```diff
@@ -79,7 +78,7 @@ class HnswIndexTest : public ::testing::Test {
     void init(bool heuristic_select_neighbors) {
         auto generator = std::make_unique<LevelGenerator>();
         level_generator = generator.get();
-        index = std::make_unique<HnswIndex>(vectors, std::make_unique<FloatSqEuclideanDistance>(),
+        index = std::make_unique<HnswIndex>(vectors, std::make_unique<SquaredEuclideanDistance>(vespalib::eval::CellType::DOUBLE),
```
Comment (havardpe) [inline searchlib/src/tests/tensor/hnsw_index/hnsw_index_test.cpp:81]: did this change from float to double?

File: searchlib/src/tests/tensor/hnsw_index/stress_hnsw_mt.cpp [hunk 3]
```diff
@@ -232,7 +231,7 @@ class Stressor : public ::testing::Test {
 
     void init() {
         uint32_t m = 16;
-        index = std::make_unique<HnswIndex>(vectors, std::make_unique<FloatSqEuclideanDistance>(),
+        index = std::make_unique<HnswIndex>(vectors, std::make_unique<SquaredEuclideanDistance>(vespalib::eval::CellType::DOUBLE),
```
Comment (havardpe) [inline searchlib/src/tests/tensor/hnsw_index/stress_hnsw_mt.cpp:234]: did this change from float to double

File: searchlib/src/vespa/searchlib/queryeval/nearest_neighbor_blueprint.cpp [hunk 4]
```diff
@@ -4,43 +4,44 @@
...
 
 namespace search::queryeval {
 
 namespace {
 
 template<typename LCT, typename RCT>
-void
-convert_cells(std::unique_ptr<Value> &original, const vespalib::eval::ValueType &want_type)
+std::unique_ptr<Value>
+convert_cells(const ValueType &new_type, TypedCells cells)
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/queryeval/nearest_neighbor_blueprint.cpp:27]: consider passing in old value as unique ptr as well

File: searchlib/src/vespa/searchlib/queryeval/nearest_neighbor_blueprint.cpp [hunk 5]
```diff
@@ -63,19 +64,23 @@ NearestNeighborBlueprint::NearestNeighborBlueprint(const queryeval::FieldSpec& f
...
     assert(_dist_fun);
     auto nns_index = _attr_tensor.nearest_neighbor_index();
     if (nns_index) {
         _dist_fun = nns_index->distance_function();
         assert(_dist_fun);
     }
+    auto query_ct = _query_tensor->cells().type;
+    CellType want_ct = _dist_fun->expected_cell_type();
+    if (query_ct != want_ct) {
+        ValueType new_type = ValueType::make_type(want_ct, _query_tensor->type().dimensions());
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/queryeval/nearest_neighbor_blueprint.cpp:79]: consider using `_query_tensor->type().cell_cast(want_ct)` here

**Review by havardpe — APPROVED** [review]

**Comment by hmusum** [comment]
FYI, There is 30-40% higher latency for 3 nearest neighbor performance tests after this PR. @aressem @baldersheim FYI

**Pull request merged**